modxcms/revolution
# Refactor Static Element File Handling

### What does it do?

Made changes and additions to modElement, select intermediate Element processors, and modMediaSource, as well as added a new system setting. These proposed changes ... 
- Improve static file handling and makes non-media source-based statics work again
- Provide the same capability of automatically creating a typed-in file name (where non-existing intermediate directories are created on-the-fly) for non-media source files (user selects "None" as source) as is available for files with an assigned media source. Note that this behavior was introduced with the implementation of the Flysystem file handling library (3.x only), which only applies to media sources.
- Create a new system setting (`static_elements_filetypes`) to enable file types specifically for static elements (ones that you wouldn't want available globally, such as php).
- Ensure that a media source’s `allowedFileTypes` property is honored (it currently is not because its value is called in the wrong way).
- Ensure that only the proper file type is allowed for Plugins and Snippets (php), as well as Chunks and Templates (anything but php, for now).
- Ensure that if a user enters a file path cannot be written or is otherwise invalid that the Element is not saved. Currently, under this condition, an error is shown regarding the invalid file but the Element gets saved anyway (yet the user is left on the create form).
- Ensure that when a static Element is duplicated, its content is duplicated as well.
- Consolidate duplicate logic into new methods in modElement.

### Why is it needed?

There have been assorted issues that have cropped up as a result of the implementation of the Flysystem file handling library that need to be fixed.

### How to test

1. Ensure all caches are completely cleared.
2. Create a few _non_ static Elements to verify correct behavior.
3. Create a variety of static Elements, both by assigning a media source and by selecting "None" as media source, by typing in a path (that does not yet exist) to verify the file gets written and is written to the proper place. You should be able to create a blank file, as well as one in which you've entered content into the Element's code field (which should save to the file).
4. Repeat step 3 above, but this time create some files (either manually or via the file browser) to verify that pointing a static element to an existing file works as expected. 
5. When running steps 3 and 4 above, at least one media source with a `basePath` set outside of the webroot should be verified as well.
6. Duplicate a few Elements to verify correct behavior.
7. Try to make the operations above fail as well, by entering invalid or duplicate data, to verify the process is stopped and validation messages are shown as expected.
8. Change the `static_elements_filetypes` setting to verify it affects the ability to use specified file types as expected.
9. Change the `allowedFileTypes` setting in one or more media sources to verify that it affects the ability to use specified file types as expected. This is the setting that will take top precedence. If blank, the global `static_elements_filetypes` setting applies and, if that setting is blank, the fallback of `upload_files` setting will be honored.

### Related issue(s)/PR(s)

Prompted by issue #15927 and related to PR #16043.
